internal/controller/customrun_controller.go

```diff
@@ -45,17 +58,118 @@ type CustomRunReconciler struct {
 // For more details, check Reconcile and its Result here:
 // - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.19.0/pkg/reconcile
 func (r *CustomRunReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	logger := log.FromContext(ctx)
 
-	// TODO(user): your logic here
+	var customrun tektonv1beta1.CustomRun
+	err := r.Get(ctx, req.NamespacedName, &customrun)
+	if apierrors.IsNotFound(err) {
+		logger.Info("reconcile: CustomRun deleted", "customrun", req.NamespacedName)
+		// Request object not found, could have been deleted after reconcile request.
+		// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+		return reconcile.Result{}, nil
+	}
+
+	if err != nil {
+		logger.Error(err, "reconcile: unable to fetch CustomRun")
+		return ctrl.Result{}, err
+	}
+
+	customSpec := customrun.Spec.CustomSpec
+	if customSpec != nil {
+		if customSpec.APIVersion == "jumpstarter.dev/v1alpha1" && customSpec.Kind == "Lease" {
+			// task already completed
+			if !customrun.Status.GetCondition(knative.ConditionSucceeded).IsUnknown() {
+				return reconcile.Result{}, err
+			}
+
+			var leaseSpec jumpstarterdevv1alpha1.LeaseSpec
+			if err := json.NewDecoder(bytes.NewBuffer(customSpec.Spec.Raw)).Decode(&leaseSpec); err != nil {
+				return reconcile.Result{}, err
+			}
+
+			var lease jumpstarterdevv1alpha1.Lease
+			err := r.Get(ctx, types.NamespacedName{
+				Namespace: customrun.Namespace,
+				Name:      customrun.Name,
+			}, &lease)
+
+			if err == nil {
+				lease.Spec = leaseSpec
+				if err := controllerutil.SetOwnerReference(&customrun, &lease, r.Scheme); err != nil {
+					return reconcile.Result{}, err
+				}
+				if err := r.Update(ctx, &lease); err != nil {
+					logger.Info("reconcile: unable to update lease", "customrun", req.NamespacedName)
+				}
+			} else if apierrors.IsNotFound(err) {
+				lease.ObjectMeta = metav1.ObjectMeta{
+					Namespace: customrun.Namespace,
```

Comment:
We need to think of the cases where the custom run is on a namespace,
and the jumpstarter namespace is on another one, to isolate concerns:
lab management <> CI runs.

Reply:
> We need to support cases where the custom run is on a namespace,
> and the jumpstarter namespace is on another one

Then we need to configure the customrun controller with a "lab mgmt namespace", since customRef does not contain a namespace field.

Reply:
Yeah, we need to figure this out. Also, we need to figure out how to do it if the Jumsptarter controller is on a different cluster than tekton.

I wonder if we should go straight into having a different controller altogether for this, then we could configure it to talk to another cluster , with its own service account to interact with jumpstarter if we need to.

Reply:
> I wonder if we should go straight into having a different controller altogether for this

Yeah this could be the best solution, so we would have a `CustomRun controller` for every namespace under which tekton jobs run, and with their own `lab namespace` configured for interacting with jumpstarter resources.